### PR TITLE
Don't run tests with custom arguments with pytest

### LINKED ISSUE #41393: Running tests with --pytest fails with unrecognized arguments: --subprocess
## 🐛 Bug

When running run_tests.py with --pytest the test test_cuda_primary_ctx.py fails with "unrecognized arguments: --subprocess"

## To Reproduce

- `python -m pytest test_cuda_primary_ctx.py --subprocess` to easily reproduce
- `python run_tests.py --pytest` for "regular" use which causes the above call

The reason is that unittest (likely) ignores all unknown parameters while pytest does not. There are 2 solutions:
- Make `run_tests.py` be aware of it and don't use pytest for such tests
- Make the test aware of pytest and/or always use pytest so the options can be added to the pytest parser

## Environment

- PyTorch 1.6.0-rc2

cc @mruberry @VitalyFedyunin

### PR DESCRIPTION
**Pull request by Flamefire**
This patch basically removes the `-m pytest` parameters when `extra_unittest_args` is used (e.g. `--subprocess`)

Fixes #41393

### CODE REVIEW

File: test/run_test.py
Comment (SsnL) [inline test/run_test.py:243]: I honestly don't think this is the correct fix.  Now you basically have the behavior that silently fails such tests. I would rather just raise if we end up in this if-statement.

Comment (Flamefire) [reply test/run_test.py:243]: My goal was to support `python run_test.py --pytest`. This currently fails due to https://github.com/pytorch/pytorch/blob/db1646e6a3a7089824ba896a14663c05390df9b2/test/run_test.py#L250 which adds `--subprocess` to `extra_unittest_args`. There might be other such calls, but didn't fully check.

For those cases it should gracefully fall back to using unittest instead of pytest because it is known that pytest doesn't support those arguments.

>  Now you basically have the behavior that silently fails such tests

I don't understand. This change is to make such tests pass. Why would they fail and even silently?

> I would rather just raise if we end up in this if-statement.

That would be an improvement as the error message would be more clear. However it doesn't help for the above described case of running all tests with `python run_test.py --pytest`

Comment (ezyang) [reply test/run_test.py:243]: The new code is brittle. It might work as the code is written today, but there are a number of reasonable changes one might make to code that calls `run_test` that would end up doing the wrong thing in those cases. If your goal is to avoid the `--subprocess` call when running with pytest, it would be better to directly modify line 250. But I am not too sure about the overall structure of this patch; maybe there is some design problem that is leading you to want to do this.

Comment (Flamefire) [reply test/run_test.py:243]: > It might work as the code is written today, but there are a number of reasonable changes one might make to code that calls run_test that would end up doing the wrong thing in those cases

Could you clarify this? I wrote the code with the intention to be general and not fail and especially not do the wrong thing. To be even more explicit I can add `and executable[-2] == '-m'`. This means this code will only ever remove `-m pytest` from the end of `executable` if it is present and any extra_unittest_args are given. When could this possibly be wrong?

> If your goal is to avoid the --subprocess call when running with pytest, it would be better to directly modify line 250.

As implied by the name `extra_unittest_args` those arguments are for unittest only, or to be precise: For the custom argument parser in the `common` submodule. Those won't ever work with pytest unless support is implemented one day. So yes I could move that down to line 250 but then it will fail for the next test that adds an argument parsed by the `common` module. So IMO this is the more general and solid solution.

> But I am not too sure about the overall structure of this patch; maybe there is some design problem that is leading you to want to do this.

Well the design problem here is that the code authors rely on the behavior that unknown arguments are ignored by the test framework. This assumption seems to hold for unittest but doesn't hold for pytest, which is kinda supported too. See comment by @mruberry https://github.com/pytorch/pytorch/issues/41393#issuecomment-659194797   
In #41393 I suggested 2 solutions: Make the tests work with pytest (likely a lot of work) or don't use pytest for subtests that are incompatible with it. In this case: All tests receiving custom arguments not known by pytest itself. This is implemented here.

Comment (mruberry) [reply test/run_test.py:243]: What about a deny list of test suites for pytest and an announcement that the test suite is skipped when calling run_test with the pytest option?

Comment (ezyang) [reply test/run_test.py:243]: @Flamefire, I want to preface that, in the grand scheme of things, these two lines don't matter too much. However, you've been writing a number of good PRs for us (thank you!) and so my goal here is more to explicate the general philosophy for how we write code in PyTorch, which may be much more helpful in the future.

> Could you clarify this? I wrote the code with the intention to be general and not fail and especially not do the wrong thing.

I agree that this code is pretty unlikely to end up blowing up in practice. It can still be brittle though! What makes this code brittle is that it assumes extra invariants on the passed in `executable` in a way that is not correlated with the documentation or ones intuitive "sense" of what a function ought to do. As you point out, you could be more explicit by testing `-m` before `pytest`; when you don't actually test for it, you have imposed the invariant "whenever there is a trailing pytest in executable, it must be preceeded by an `-m`". Imagine writing that in the documentation for this function! That would be a pretty nutty thing to write. Even if you add the test for `-m`, now the invariant for executable is something like, "any executable that could be able to run the test in question, but if you intend to run the program with pytest, it must be done in such a way so that the last two arguments are `-m pytest`" which is still a pretty nutty invariant to require. It is true: the invariant is unlikely to be broken. But if we're thinking about understandability of the code, whether or not the invariants are true or not isn't enough: they should also be understandable!

While rereviewing this discussion, I also realized that SsnL's problem is subtly different than mine.

> I don't understand. This change is to make such tests pass. Why would they fail and even silently?

Simon's objection, if I understand correctly, is this: if I pass `--pytest` to `run_tests.py`, I might expect the semantics of this flag to be "use pytest on all of the tests", not "make a best effort attempt to use pytest and fall back to stock python runner if it doesn't work". This distinction could matter if, e.g., I want to use pytest because there is a specific feature in pytest I need to exercise.

What are the true semantics? Well, typically, you would defer to the original author of the code. In that case, the author is @SsnL He added it in #13416.

OK, so what should you do? I think @mruberry has made a reasonable suggestion, but let me make another suggestion of my own: why don't we push the call to `get_executable_command` into `run_test` (perhaps even inlining it)? Based on my reading of the code, the only reason we thread executable around is so that the MPI logic can add a prefix to the command for mpirun. In that case, let's just make `run_test` take an (optional) `executable_prefix` argument.

Once you do this refactor, now you can fix things in a very logical way: inside `get_executable_command` you can test if you are going to go the pytest route, and then you can either raise an error or suppress the flag (depending on what @SsnL says) when there would be an incompatibility. Now you don't have to test for pytest in executable; you can directly express the intended semantics.

Hope this helps.

Comment (Flamefire) [reply test/run_test.py:243]: Ok, thanks for the clarification. My motivation was to solve the issue (which seems to affect a single test only at this point, but I might be wrong) in a generic way while using the minimally invasive variant. Of course this is up for discussion as that is what PR reviews are for.
I don't think a deny list would work as that is yet another list which needs to be kept up to date without CI checking this (I assume, as otherwise the issue would have been found earlier)
Your suggestion makes a lot of sense. Let me propose a full change:
- refactor `run_test` to take `launcher_cmd` as an optional parameter with default set empty (I'd prefer that name over `executable_prefix` as I think it makes the intention clearer and `launcher` is (IIRC) what the mpirun cmd is called elsewhere too)
- move `get_executable_command` above `run_test` (I'd keep it but move it close to usage site)
- Add an extra parameter `allow_pytest` to `get_executable_command` and set that to `len(extra_unittest_args) == 0` in `run_test`
- Print a message to stdout when pytest was requested but is not used from inside `get_executable_command`
- Finally replace the current `Running <testname>` message by doing it from within `run_test` and show the full cmdline used like: `Running <testname>: <cmdline>`, this would have helped at least me a couple times already and also makes this clear

Comment (ezyang) [reply test/run_test.py:243]: sgtm. Some minor nits:

> Print a message to stdout when pytest was requested but is not used from inside get_executable_command

better to print it to stderr. I think Simon would prefer it if we raised an error but in the spirit of moving things along I'll say I'm ok with this new behavior.

Comment (Flamefire) [reply test/run_test.py:243]: Sure, can do. I'd very much NOT use an error because my naive attempt was `python run_test.py --pytest` which failed and IMO it is better to have that fall back to unittest instead of failing. Sure it's a compromise, as usual.

May I understand your message as approval of the proposed changes (with stderr) or shall I wait for Simon to confirm?

Comment (ezyang) [reply test/run_test.py:243]: If you really want to wait, sure, but I'll approve it and land it if you go ahead too.

Comment (Flamefire) [reply test/run_test.py]: I implemented the proposed changes and rebased my branch to avoid conflicts. Only thing I changed is the last point: I didn't remove the original print of `Running <testname>` but added one just before command execution. Reason is that the original print shows the test to be run which might be different to the actual command. I.e. it seems the `test` can contain a dot, I guess some legacy support for `TestModule.MyTest` stuff although the 2nd part is ignored and it looks superflous to me as the given `test` (which is passed via cmdline) cannot contain a dot as it must match the entries in `TESTS`

Anyway I'll leave that for later.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
